// @nuintun/qrcode/esm/qrcode/encoder/BitBuffer.js
/**
 * @module QRCode
 * @package @nuintun/qrcode
 * @license MIT
 * @version 3.3.0
 * @author nuintun <nuintun@qq.com>
 * @description A pure JavaScript QRCode encode and decode library.
 * @see https://github.com/nuintun/qrcode#readme
 */

/**
 * @module BitBuffer
 * @author nuintun
 * @author Kazuhiko Arase
 */
var BitBuffer = /*#__PURE__*/ (function () {
  function BitBuffer() {
    this.length = 0;
    this.buffer = [];
  }
  BitBuffer.prototype.getBuffer = function () {
    return this.buffer;
  };
  BitBuffer.prototype.getLengthInBits = function () {
    return this.length;
  };
  BitBuffer.prototype.getBit = function (index) {
    return ((this.buffer[(index / 8) >> 0] >>> (7 - (index % 8))) & 1) === 1;
  };
  BitBuffer.prototype.put = function (num, length) {
    for (var i = 0; i < length; i++) {
      this.putBit(((num >>> (length - i - 1)) & 1) === 1);
    }
  };
  BitBuffer.prototype.putBit = function (bit) {
    var buffer = this.buffer;
    if (this.length === buffer.length * 8) {
      buffer.push(0);
    }
    if (bit) {
      buffer[(this.length / 8) >> 0] |= 0x80 >>> this.length % 8;
    }
    this.length++;
  };
  return BitBuffer;
})();

export { BitBuffer };
